src/common/containers/NewSet.tsx

```diff
@@ -0,0 +1,221 @@
+
+import * as w from '../../common/types';
+import { id as generateId } from '../util/common';
+import { getDisplayedCards } from '../util/cards';
+import ActiveDeck from '../components/cards/ActiveDeck';
+import CardCollection from '../components/cards/CardCollection';
+import EnergyCurve from '../components/cards/EnergyCurve';
+import FilterControls from '../components/cards/FilterControls';
+import LayoutControls from '../components/cards/LayoutControls';
+import SearchControls from '../components/cards/SearchControls';
+import SortControls from '../components/cards/SortControls';
+import * as collectionActions from '../actions/collection';
+
+interface NewSetStateProps {
+  setBeingEdited: w.Set | null
+  allCards: w.CardInStore[]
+  user: fb.User | null
+}
+
+interface NewSetDispatchProps {
+  onSaveSet: (set: w.Set) => void
+}
+
+type NewSetProps = NewSetStateProps & NewSetDispatchProps & { history: History };
+
+interface NewSetState {
+  filters: {
+    robots: boolean
+    events: boolean
+    structures: boolean
+  }
+  costRange: [number, number]
+  sortCriteria: 0 | 1 | 2 | 3 | 4 | 5 | 6
+  sortOrder: 0 | 1
+  searchText: string
+  selectedCardIds: string[]
+  layout: 0 | 1
+}
+
+function mapStateToProps(state: w.State): NewSetStateProps {
+  return {
+    setBeingEdited: state.collection.setBeingEdited,
+    allCards: state.collection.cards,
+    user: state.global.user
+  };
+}
+
+function mapDispatchToProps(dispatch: Dispatch<AnyAction>): NewSetDispatchProps {
+  return {
+    onSaveSet: (set: w.Set) => {
+      dispatch(collectionActions.saveSet(set));
+    }
+  };
+}
+
+/**
+ * Container encapsulating display and logic for creating and editing Sets.
+ * @TODO Reduce duplication between NewSet and Deck containers.
+ */
+class NewSet extends React.Component<NewSetProps, NewSetState> {
+  public state: NewSetState = {
+    filters: {
+      robots: true,
+      events: true,
+      structures: true
+    },
+    costRange: [0, 20],
+    sortCriteria: 3,
+    sortOrder: 0,
+    searchText: '',
+    selectedCardIds: [],
+    layout: 0
+  };
+
+  constructor(props: NewSetProps) {
+    super(props);
+
+    if (props.setBeingEdited) {
+      this.setState({ selectedCardIds: props.setBeingEdited.cards.map((c) => c.id) });
+    }
+  }
+
+  get selectedCards(): w.CardInStore[] {
+    return compact(this.state.selectedCardIds.map((id) => find(this.props.allCards, { id })));
+  }
+
+  get displayedCards(): w.CardInStore[] {
+    const { searchText, filters, costRange, sortCriteria, sortOrder } = this.state;
+    return getDisplayedCards(this.props.allCards, { searchText, filters, costRange, sortCriteria, sortOrder });
+  }
+
+  public render(): JSX.Element {
+    const { setBeingEdited, user } = this.props;
+    const { layout, selectedCardIds } = this.state;
+
+    return (
+      <div>
+        <Helmet title={setBeingEdited ? "Editing Set" : "Creating Set"} />
+
+        <div style={{
+          display: 'flex',
```

Comment:
If this component is a model for future Typescript React components, lets pull out the styles into a separate object and use `withStyles`